File: pkg/grpc/solver/target.go
package solver

import (
	"fmt"
	"strconv"
	"strings"
)

// Target holds the parsed dominion grpc target parts.
type Target struct {
	App     string
	Service string
	Port    int
}

// ParseTarget parses app/service:port style dominion targets.
func ParseTarget(raw string) (*Target, error) {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return nil, fmt.Errorf("invalid target %q: want app/service:port", raw)
	}

	if withoutScheme, ok := strings.CutPrefix(trimmed, Scheme+":///"); ok {
		trimmed = withoutScheme
	} else if strings.Contains(trimmed, "://") {
		return nil, fmt.Errorf("invalid target %q: want app/service:port or %s:///app/service:port", raw, Scheme)
	}

	parts := strings.SplitN(trimmed, ":", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return nil, fmt.Errorf("invalid target %q: want app/service:port", raw)
	}
	if strings.Contains(parts[1], ":") {
		return nil, fmt.Errorf("invalid target %q: want app/service:port", raw)
	}

	path := strings.SplitN(parts[0], "/", 2)
	if len(path) != 2 || path[0] == "" || path[1] == "" || strings.Contains(path[1], "/") {
		return nil, fmt.Errorf("invalid target %q: want app/service:port", raw)
	}

	for _, ch := range parts[1] {
		if ch < '0' || ch > '9' {
			return nil, fmt.Errorf("invalid target %q: port must be numeric", raw)
		}
	}

	port, err := strconv.Atoi(parts[1])
	if err != nil || port < 1 || port > 65535 {
		return nil, fmt.Errorf("invalid target %q: want app/service:port", raw)
	}

	return &Target{
		App:     path[0],
		Service: path[1],
		Port:    port,
	}, nil
}

File: common/js/resolver/src/target.ts
import { InvalidTargetError } from "./errors.js";

/**
 * Numeric port selector — port must be an integer in `0..65535`.
 */
export type PortSelector = { kind: "number"; port: number };

/**
 * Parsed dominion service target.
 */
export interface Target {
  app: string;
  service: string;
  port: PortSelector;
}

const DOMINION_SCHEME_PREFIX = "dominion:///";
const MAX_PORT = 65535;

/**
 * Parses a dominion target string of the form `app/service:port`.
 *
 * Accepts:
 * - Direct format: `app/service:port`
 * - Dominion scheme: `dominion:///app/service:port`
 *
 * The port must be a non-negative integer in the range 0–65535.
 * Named ports (e.g. `:grpc`) are not supported — callers must
 * resolve the port number themselves before constructing the target.
 *
 * Throws `InvalidTargetError` for invalid formats.
 */
export function parseTarget(raw: string): Target {
  let trimmed = raw.trim();
  if (trimmed === "") {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: want app/service:port`,
    );
  }

  if (trimmed.startsWith(DOMINION_SCHEME_PREFIX)) {
    trimmed = trimmed.slice(DOMINION_SCHEME_PREFIX.length);
  } else if (trimmed.includes("://")) {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: want app/service:port or dominion:///app/service:port`,
    );
  }

  const colonIdx = trimmed.indexOf(":");
  let pathPart: string;
  let portPart: string | undefined;
  let hasPort = false;

  if (colonIdx !== -1) {
    pathPart = trimmed.slice(0, colonIdx);
    portPart = trimmed.slice(colonIdx + 1);
    hasPort = true;
  } else {
    pathPart = trimmed;
  }

  const slashIdx = pathPart.indexOf("/");
  if (slashIdx === -1) {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: want app/service:port`,
    );
  }

  const appPart = pathPart.slice(0, slashIdx).trim();
  const servicePart = pathPart.slice(slashIdx + 1).trim();
  const trimmedPortPart = (portPart ?? "").trim();

  if (appPart === "" || servicePart === "" || servicePart.includes("/")) {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: want app/service:port`,
    );
  }

  if (trimmedPortPart === "") {
    if (hasPort) {
      throw new InvalidTargetError(
        `invalid target ${JSON.stringify(raw)}: port must be numeric`,
      );
    }
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: port is required`,
    );
  }

  if (!/^\d+$/.test(trimmedPortPart)) {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: port must be numeric, got ${JSON.stringify(trimmedPortPart)}`,
    );
  }

  const numericPort = parseInt(trimmedPortPart, 10);
  if (numericPort < 0 || numericPort > MAX_PORT) {
    throw new InvalidTargetError(
      `invalid target ${JSON.stringify(raw)}: port out of range`,
    );
  }

  return {
    app: appPart,
    service: servicePart,
    port: { kind: "number", port: numericPort },
  };
}
